core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -224,23 +231,35 @@ public synchronized void schedulingComplete()
             return;
         }
 
-        if (isFlushing()) {
-            stateMachine.transitionToFlushing();
+        try {
+            allTasksLock.readLock().lock();
+            if (isFlushing(flushingOrFinishedTasksCounter.get())) {
+                stateMachine.transitionToFlushing();
+            }
+            if (isFinished(finishedTasksCounter.get())) {
+                stateMachine.transitionToFinished();
+            }
         }
-        if (finishedTasks.containsAll(allTasks)) {
-            stateMachine.transitionToFinished();
+        finally {
+            allTasksLock.readLock().unlock();
         }
 
         for (PlanNodeId partitionedSource : stage.getFragment().getPartitionedSources()) {
             schedulingComplete(partitionedSource);
         }
     }
 
-    private synchronized boolean isFlushing()
+    private boolean isFlushing(long flushingOrFinishedTasks)
```

Comment:
annotate with `@GuardedBy("allTasksLock")`. Otherwise, you get intellij warnings like this:

![image](https://user-images.githubusercontent.com/890111/190411714-f17b5258-20b4-4245-90bc-d74b9d8dedb7.png)

Enable those if you have those disabled btw.

Reply:
Thanks